src/main/scala/nl.knaw.dans.easy.deposit/StateManager.scala

```diff
@@ -0,0 +1,110 @@
+
+import scala.util.{ Failure, Success, Try }
+
+case class StateManager(depositDir: File, submitBase: File) {
+  private val stateDescriptionKey = "state.description"
+  private val stateLabelKey = "state.label"
+  private val bagIdKey = "bag-store.bag-id"
+  val draftProps = new PropertiesConfiguration(
+    (depositDir / "deposit.properties").toJava
+  )
+  private lazy val submittedProps = new PropertiesConfiguration(
+    (submitBase / getProp(bagIdKey) / "deposit.properties").toJava
+  )
+
+  def getStateInfo: Try[StateInfo] = Try {
+    val draftState = getStateLabel()
+    draftState match {
+      case State.submitted | State.inProgress =>
+        val newState: StateInfo = getProp(stateLabelKey, submittedProps) match {
+          case "SUBMITTED" => StateInfo(State.submitted, getStateDescription())
+          case "REJECTED" => StateInfo(State.rejected, getStateDescription(submittedProps))
+          case "FEDORA_ARCHIVED" => StateInfo(State.archived, "The dataset is published in https://easy.dans.knaw.nl/ui")
+          case "IN_REVIEW" | "FAILED" => StateInfo(State.inProgress, "The dataset is visible for you under your datasets in https://easy.dans.knaw.nl/ui")
+          case str: String => throw InvalidPropertyException(stateLabelKey, str, submittedProps)
+        }
+        saveNewState(newState)
+      case State.draft | State.rejected | State.archived =>
+        StateInfo(draftState, getStateDescription())
+    }
+  }
+
+  def canChangeState(newStateInfo: StateInfo): Try[Unit] = getStateInfo.flatMap { oldStateInfo =>
+    val oldState = oldStateInfo.state
+    val newState = newStateInfo.state
+    (oldState, newState) match {
+      case (State.draft, State.submitted) |
+           (State.rejected, State.draft) => Success(())
+      case _ => Failure(IllegalStateTransitionException(oldState, newState))
+    }
+  }
+
+  /** @return bag-store.bag-id in case the new state is submitted */
+  def changeState(newStateInfo: StateInfo): Try[UUID] = getStateInfo.flatMap { old =>
+    // getStateInfo has been called by canChangeState, but it is not an IO action so no optimisation
+    (old.state, newStateInfo.state) match {
+      case (State.draft, State.submitted) =>
+        val bagStoreBagId = UUID.randomUUID()
+        draftProps.setProperty(bagIdKey, bagStoreBagId)
+        saveNewState(newStateInfo)
+        Success(bagStoreBagId)
+      case (State.rejected, State.draft) =>
+        draftProps.clearProperty(bagIdKey)
+        saveNewState(newStateInfo)
+        Success(null)
+      case (oldState, newState) =>
+        Failure(IllegalStateTransitionException(oldState, newState))
+    }
+  }
+
+  private def saveNewState(newStateInfo: StateInfo) = {
+    draftProps.setProperty(stateLabelKey, newStateInfo.state.toString)
+    draftProps.setProperty(stateDescriptionKey, newStateInfo.stateDescription)
+    draftProps.save()
+    newStateInfo
+  }
+
+  private def getStateLabel(props: PropertiesConfiguration = draftProps): State = {
```

Comment:
I wonder why you start throwing exceptions here and in the methods below. Why not use proper `Try` wrappers?

Reply:
It would get complicated to handle the results of `getStateLabel` called outside  and `getStateDescription` called inside the pattern match of `getStateInfo`.